File: java/src/main/java/com/phenixrts/edgeauth/EdgeAuth.java
/**
 * Copyright 2019 Phenix Real Time Solutions, Inc. All Rights Reserved.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.phenixrts.edgeauth;

import java.util.Date;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;

public class EdgeAuth {
  private static final long DEFAULT_EXPIRATION_IN_SECONDS = 3600;

  /**
   * Edge Auth token generator CLI.
   *
   * @param args the CLI arguments
   */
  public static void main(String[] args) {
    final Options options = createOptions();
    final CommandLineParser parser = new DefaultParser();
    final CommandLine cmd;

    try {
      cmd = parser.parse(options, args);
    } catch (ParseException e) {
      HelpFormatter formatter = new HelpFormatter();
      formatter.printHelp(EdgeAuth.class.getSimpleName(), options);

      System.exit(1);

      return;
    }

    if (cmd.hasOption("help")) {
      HelpFormatter formatter = new HelpFormatter();
      formatter.printHelp(EdgeAuth.class.getSimpleName(), options);

      return;
    }

    final TokenBuilder tokenBuilder = new TokenBuilder()
        .withApplicationId(cmd.getOptionValue("applicationId"))
        .withSecret(cmd.getOptionValue("secret"));

    if (cmd.hasOption("uri")) {
      tokenBuilder.withUri(cmd.getOptionValue("uri"));
    }

    if (cmd.hasOption("expiresAt")) {
      tokenBuilder.expiresAt(new Date(Long.parseLong(cmd.getOptionValue("expiresAt"), 10)));
    } else {
      tokenBuilder.expiresInSeconds(Long.parseLong(cmd.getOptionValue("expiresInSeconds", Long.toString(DEFAULT_EXPIRATION_IN_SECONDS)), 10));
    }

    if (cmd.hasOption("authenticationOnly")) {
      tokenBuilder.forAuthenticationOnly();
    }

    if (cmd.hasOption("streamingOnly")) {
      tokenBuilder.forStreamingOnly();
    }

    if (cmd.hasOption("publishingOnly")) {
      tokenBuilder.forPublishingOnly();
    }

    if (cmd.hasOption("capability")) {
      for (final String capability: cmd.getOptionValues("capability")) {
        tokenBuilder.withCapability(capability);
      }
    }

    if (cmd.hasOption("sessionId")) {
      tokenBuilder.forSession(cmd.getOptionValue("sessionId"));
    }

    if (cmd.hasOption("remoteAddress")) {
      tokenBuilder.forRemoteAddress(cmd.getOptionValue("remoteAddress"));
    }

    if (cmd.hasOption("originStreamId")) {
      tokenBuilder.forOriginStream(cmd.getOptionValue("originStreamId"));
    }

    if (cmd.hasOption("channel")) {
      tokenBuilder.forChannel(cmd.getOptionValue("channel"));
    }

    if (cmd.hasOption("channelAlias")) {
      tokenBuilder.forChannelAlias(cmd.getOptionValue("channelAlias"));
    }

    if (cmd.hasOption("room")) {
      tokenBuilder.forRoom(cmd.getOptionValue("room"));
    }

    if (cmd.hasOption("roomAlias")) {
      tokenBuilder.forRoomAlias(cmd.getOptionValue("roomAlias"));
    }

    if (cmd.hasOption("tag")) {
      tokenBuilder.forTag(cmd.getOptionValue("tag"));
    }

    if (cmd.hasOption("applyTag")) {
      tokenBuilder.applyTag(cmd.getOptionValue("applyTag"));
    }

    final String token;
    final String tokenObjectJson;
    try {
      token = tokenBuilder.build();
      tokenObjectJson = tokenBuilder.getValue();
    } catch (Exception e) {
      System.err.println(e.getMessage());
      System.exit(7);

      return;
    }

    System.out.println(tokenObjectJson);
    System.out.println(token);
  }

  private static Options createOptions() {
    final Options options = new Options();

    options.addOption("y", "uri", true, "The backend URI");
    options.addOption("u", "applicationId", true, "The application ID");
    options.addOption("w", "secret", true, "The application secret");
    options.addOption("l", "expiresInSeconds", true, "Token life time in seconds");
    options.addOption("e", "expiresAt", true, "Token expires at timestamp measured in milliseconds since UNIX epoch");
    options.addOption("a", "authenticationOnly", false, "Token can be used for authentication only");
    options.addOption("s", "streamingOnly", false, "Token can be used for streaming only");
    options.addOption("p", "publishingOnly", false, "Token can be used for publishing only");
    options.addOption("b", "capability", true, "Comma separated list of capabilities, e.g. for publishing");
    options.addOption("z", "sessionId", true, "Token is limited to the given session");
    options.addOption("x", "remoteAddress", true, "Token is limited to the given remote address");
    options.addOption("o", "originStreamId", true, "[STREAMING] Token is limited to the given origin stream");
    options.addOption("c", "channel", true, "[STREAMING] Token is limited to the given channel");
    options.addOption("i", "channelAlias", true, "[STREAMING] Token is limited to the given channel alias");
    options.addOption("m", "room", true, "[STREAMING] Token is limited to the given room");
    options.addOption("n", "roomAlias", true, "[STREAMING] Token is limited to the given room alias");
    options.addOption("t", "tag", true, "[STREAMING] Token is limited to the given origin stream tag");
    options.addOption("r", "applyTag", true, "[REPORTING] Apply tag to the new stream");
    options.addOption("h", "help", false, "Print this message");

    return options;
  }
}

File: dotnet/PhenixRTS.EdgeAuth.CLI/EdgeAuth.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Utility.CommandLine;

namespace PhenixRTS.EdgeAuth.CLI
{
    public class EdgeAuth
    {
        [Argument('y', "uri", "The backend URI")]
        private static string Uri { get; set; }

        [Argument('u', "applicationId", "The application ID")]
        private static string ApplicationId { get; set; }

        [Argument('w', "secret", "The application secret")]
        private static string Secret { get; set; }

        [Argument('l', "expiresInSeconds", "Token life time in seconds")]
        private static string ExpiresInSeconds { get; set; }

        [Argument('e', "expiresAt", "Token expires at timestamp measured in milliseconds since UNIX epoch")]
        private static string ExpiresAt { get; set; }

        [Argument('a', "authenticationOnly", "Token can be used for authentication only")]
        private static string AuthenticationOnly { get; set; }

        [Argument('s', "streamingOnly", "Token can be used for streaming only")]
        private static string StreamingOnly { get; set; }

        [Argument('p', "publishingOnly", "Token can be used for publishing only")]
        private static string PublishingOnly { get; set; }

        [Argument('b', "capability", "Comma separated list of capabilities, e.g. for publishing")]
        private static string Capability { get; set; }

        [Argument('z', "sessionId", "Token is limited to the given session")]
        private static string SessionId { get; set; }

        [Argument('x', "remoteAddress", "Token is limited to the given remote address")]
        private static string RemoteAddress { get; set; }

        [Argument('o', "originStreamId", "[STREAMING] Token is limited to the given origin stream")]
        private static string OriginStreamId { get; set; }

        [Argument('c', "channel", "[STREAMING] Token is limited to the given channel")]
        private static string Channel { get; set; }

        [Argument('i', "channelAlias", "[STREAMING] Token is limited to the given channel alias")]
        private static string ChannelAlias { get; set; }

        [Argument('m', "room", "[STREAMING] Token is limited to the given room")]
        private static string Room { get; set; }

        [Argument('n', "roomAlias", "[STREAMING] Token is limited to the given room alias")]
        private static string RoomAlias { get; set; }

        [Argument('t', "tag", "[STREAMING] Token is limited to the given origin stream tag")]
        private static string Tag { get; set; }

        [Argument('r', "applyTag", "[REPORTING] Apply tag to the new stream")]
        private static string ApplyTag { get; set; }

        [Argument('h', "help", "Print this message")]
        private static string Help { get; set; }

        private const long DEFAULT_EXPIRATION_IN_SECONDS = 3600;

        /// <summary>
        /// Edge Auth token generator CLI.
        /// </summary>
        /// <param name="args">Args the CLI arguments</param>
        static void Main(string[] args)
        {
            try
            {
                Arguments.Populate();
                IEnumerable<ArgumentInfo> arguments = Arguments.GetArgumentInfo();
                CheckArguments(args, arguments);

                if (Help != null)
                {
                    PrintHelp(arguments);
                    return;
                }

                if (string.IsNullOrEmpty(ApplicationId) || string.IsNullOrEmpty(Secret))
                {
                    Console.WriteLine("You must provide both \"applicationId\" and \"secret\"");
                    return;
                }

                TokenBuilder tokenBuilder = new TokenBuilder().WithApplicationId(ApplicationId)
                                                              .WithSecret(Secret);

                BuildToken(tokenBuilder);

                try
                {
                    string tokenObjectJson = tokenBuilder.GetValue();
                    Console.WriteLine(tokenObjectJson);

                    string token = tokenBuilder.Build();
                    Console.WriteLine(token);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.Message);
                    return;
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }
        }

        private static void BuildToken(TokenBuilder tokenBuilder)
        {
            if (!string.IsNullOrEmpty(Uri))
            {
                tokenBuilder.WithUri(Uri);
            }

            if (!string.IsNullOrEmpty(ExpiresAt))
            {
                tokenBuilder.ExpiresAt(DateTimeOffset.FromUnixTimeMilliseconds(Convert.ToInt64(ExpiresAt, 10)).UtcDateTime);
            }
            else
            {
                long expirationInSeconds = DEFAULT_EXPIRATION_IN_SECONDS;

                if (!string.IsNullOrEmpty(ExpiresInSeconds))
                {
                    expirationInSeconds = Convert.ToInt64(ExpiresInSeconds, 10);
                }

                tokenBuilder.ExpiresInSeconds(expirationInSeconds);
            }

            if (AuthenticationOnly != null)
            {
                tokenBuilder.ForAuthenticationOnly();
            }

            if (StreamingOnly != null)
            {
                tokenBuilder.ForStreamingOnly();
            }

            if (PublishingOnly != null)
            {
                tokenBuilder.ForPublishingOnly();
            }

            if (!string.IsNullOrEmpty(Capability))
            {
                foreach (string capability in Capability.Split(','))
                {
                    tokenBuilder.WithCapability(capability.Trim());
                }
            }

            if (!string.IsNullOrEmpty(SessionId))
            {
                tokenBuilder.ForSession(SessionId);
            }

            if (!string.IsNullOrEmpty(RemoteAddress))
            {
                tokenBuilder.ForRemoteAddress(RemoteAddress);
            }

            if (!string.IsNullOrEmpty(OriginStreamId))
            {
                tokenBuilder.ForOriginStream(OriginStreamId);
            }

            if (!string.IsNullOrEmpty(Channel))
            {
                tokenBuilder.ForChannel(Channel);
            }

            if (!string.IsNullOrEmpty(ChannelAlias))
            {
                tokenBuilder.ForChannelAlias(ChannelAlias);
            }

            if (!string.IsNullOrEmpty(Room))
            {
                tokenBuilder.ForRoom(Room);
            }

            if (!string.IsNullOrEmpty(RoomAlias))
            {
                tokenBuilder.ForRoomAlias(RoomAlias);
            }

            if (!string.IsNullOrEmpty(Tag))
            {
                tokenBuilder.ForTag(Tag);
            }

            if (!string.IsNullOrEmpty(ApplyTag))
            {
                tokenBuilder.ApplyTag(ApplyTag);
            }
        }

        private static void CheckArguments(string[] args, IEnumerable<ArgumentInfo> arguments)
        {
            for (int i = 0; i < args.Length; i++)
            {
                string arg = args[i];
                if (!arguments.Any(a => $"-{a.ShortName}" == arg || $"--{a.LongName}" == arg) &&
                    arg != "--" &&
                    arg != "-"
                    && (arg.StartsWith("-") || arg.StartsWith("--")))
                {
                    throw new Exception("unknown option " + arg);
                }
            }
        }

        private static void PrintHelp(IEnumerable<ArgumentInfo> arguments)
        {
            foreach (ArgumentInfo argument in arguments)
            {
                Console.WriteLine(" Short: -{0}, Long: --{1} - {2}", argument.ShortName, argument.LongName, argument.HelpText);
            }
        }
    }
}
